src/subdataframe/subdataframe.jl

```diff
@@ -176,32 +176,106 @@ Base.@propagate_inbounds Base.getindex(df::SubDataFrame, row_ind::typeof(!),
                                        col_inds::MultiColumnIndex) =
     select(df, col_inds, copycols=false)
 
-
 Base.@propagate_inbounds function Base.setindex!(sdf::SubDataFrame, val::Any, idx::CartesianIndex{2})
-    setindex!(sdf, val, idx[1], idx[2])
+    return setindex!(sdf, val, idx[1], idx[2])
 end
+
 Base.@propagate_inbounds function Base.setindex!(sdf::SubDataFrame, val::Any, ::Colon, colinds::Any)
-    parent(sdf)[rows(sdf), parentcols(index(sdf), colinds)] = val
+    if colinds isa SymbolOrString && columnindex(sdf, colinds) == 0
+        if !is_column_adding_allowed(sdf)
+            throw(ArgumentError("creating new columns in a SubDataFrame that subsets " *
+                                "columns of its parent data frame is disallowed"))
+        end
+        if !(val isa AbstractVector && nrow(sdf) == length(val))
+            throw(ArgumentError("Assigned value must be a vector with length " *
+                                "equal to number of rows in the SubDataFrame"))
+        end
+        T = eltype(val)
+        newcol = Tables.allocatecolumn(Union{T, Missing}, nrow(parent(sdf)))
+        fill!(newcol, missing)
```

Comment:
`similar` is even better when we have a reference column to use:
```suggestion
        newcol = similar(old_col, promote_type(T, S))
        copyto!(newcol, old_col)
```

Reply:
Changed to `similar`, but you probably copy-pasted wrong code here.